API Refactor

drGrove opened the pull request:
Migrate API from a single endpoint to a true restful interface.

* Add Auth Handler to allow for more than just PGP signature
* Separate endpoints for certificate generation and user addition/removal
* Sign entire payload with PGP Signatures
* Move PGP Signature out of body and into header as base64 encoded string
* Remove clear signing from expected input
* Break out integration tests from unit tests
* Remove handler middleware to simplify code paths
* Signature timestamp is checked and must be within a threadshold to prevent replay attacks

Breaking Changes:

* API interface has totally changed from using / to /certs and /users.
* PGP Signatures are no longer in the body of the request and no longer just sign
a simple string but instead the entire JSON payload. This now means that signed
messages will require keys within the JSON document to be sorted as pythons
json.loads sorts keys by default for validating signatures

drGrove committed:
Start refactor

drGrove committed:
API Rewrite

Migrate API from a single endpoint to a true restful interface.

* Add Auth Handler to allow for more than just PGP signature
* Separate endpoints for certificate generation and user addition/removal
* Sign entire payload with PGP Signatures
* Move PGP Signature out of body and into header as base64 encoded string
* Remove clear signing from expected input
* Break out integration tests from unit tests
* Remove handler middleware to simplify code paths
* Signature timestamp is checked and must be within a threadshold to prevent replay attacks

Breaking Changes:

* API interface has totally changed from using / to /certs and /users.
* PGP Signatures are no longer in the body of the request and no longer just sign
a simple string but instead the entire JSON payload. This now means that signed
messages will require keys within the JSON document to be sorted as pythons
json.loads sorts keys by default for validating signatures

james-callahan commented on Makefile:
```diff
@@ -80,6 +80,22 @@ endif
+	$(MAKE) stop-postgres
+endif
+
+.PHONY: test.dev
```
```suggestion
.PHONY: integration-test.dev
```

james-callahan commented on Makefile:
```diff
@@ -80,6 +80,22 @@ endif
+.PHONY: integration-test
+integration-test:
+ifeq "${CI}" ""
+	-$(MAKE) stop-postgres
+	$(MAKE) run-postgres
```
Why restart the database instead of e.g. dropping the schema?

drGrove replied:
This ensures all the migrations run

james-callahan replied:
So would dropping the schema...

james-callahan commented on integration_test/test_other.py:
```diff
@@ -0,0 +1,59 @@
+import unittest
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
```
unused

james-callahan commented on integration_test/test_users.py:
```diff
@@ -0,0 +1,382 @@
+from integration_test.base import SQLiteBaseTestCase
+
+
+# logging.disable(logging.CRITICAL)
```
?

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,93 @@
+                data
+            )
+            if verified.trust_level is not None and verified.trust_level >= verified.TRUST_ULTIMATE:
+                logger.debug(f"autheticated user {verified.pubkey_fingerprint} is admin")
```
```suggestion
                logger.debug(f"authenticated user {verified.pubkey_fingerprint} is admin")
```

james-callahan commented on mtls_server/server.py:
```diff
@@ -72,13 +101,178 @@ def get_ca_cert():
+        csr = cert_processor.get_csr(csr_str)
+        if csr is None:
+            return error_response("Could not load CSR")
+        cert = None
```
Shouldn't there be a lot more validation of the `csr` here?

drGrove replied:
https://github.com/drGrove/mtls-server/blob/refactor/api/mtls_server/cert_processor.py#L446

https://github.com/drGrove/mtls-server/blob/refactor/api/mtls_server/cert_processor.py#L452

james-callahan replied:
What http status code would those validation failures lead to?

drGrove replied:
501 is the default

james-callahan replied:
validation failures should be a 4xx error; not a 501.

james-callahan commented on mtls_server/utils.py:
```diff
@@ -185,3 +195,33 @@ def import_and_trust(key_data, gpg):
     if not os.path.isdir(path):
         os.makedirs(path)
+
+def time_in_range(start, end, t):
```
This only seems to have a single call site. I'd inline it.

james-callahan commented on mtls_server/utils.py:
```diff
@@ -185,3 +195,33 @@ def import_and_trust(key_data, gpg):
+
+def add_and_trust_user(gpg, fingerprint, keyserver="keyserver.ubuntu.com"):
+    logger.info(f"Retrieving key {fingerprint} from {keyserver}")
+    result = gpg.recv_keys(
+        keyserver,
+        fingerprint,
+    )
```
Does this imply that if you're not on the keyserver you can't be added?

drGrove replied:
Currently yes

drGrove replied:
This has been a known issue, you either need to add them to the keyring manually, or include them in the seed folder.

We'll need to make an issue to allow for uploading keys are part of the post request.

drGrove committed:
Add integration tests to drone

drGrove committed:
Fix create-ca script to pass in correct attributes to CertProcessor

drGrove committed:
Remove unused imports

drGrove committed:
Disable logging on test runs to make output cleaner

james-callahan commented on integration_test/test_users.py:
```diff
@@ -0,0 +1,382 @@
+from integration_test.base import SQLiteBaseTestCase
+
+
+logging.disable(logging.CRITICAL)
```
Shouln't this be before/after tests? otherwise potential bugs/issues with the test framework itself would be lost.

james-callahan commented on mtls_server/server.py:
```diff
@@ -72,13 +101,178 @@ def get_ca_cert():
+        body = request.get_json()
+        logger.debug(body)
+        if not body:
+            return error_response(f"Could not parse body", 400)
```
Why use `f` strings instead of normal strings?

james-callahan commented on mtls_server/utils.py:
```diff
@@ -185,3 +195,33 @@ def import_and_trust(key_data, gpg):
+
+def has_user(gpg, fingerprint):
+    keys = gpg.list_keys(keys=fingerprint)
+    if len(keys) == 0:
+        return False
+    return True
```
```suggestion
    return len(keys) != 0
```

drGrove committed:
Fix spelling issue on authenticated log

drGrove committed:
Fix phony for integration-test.dev

drGrove committed:
Move logging enable/disable to the test setup/teardown to prevent logs from test framework being lost

drGrove committed:
Remove fstring from non-templated strings in error messages

drGrove committed:
More consise return for has_user

drGrove committed:
Return 4xx error when user provided information causes error

drGrove committed:
Allow for multiple coverage reports to be made and merged for coveralls

drGrove committed:
Add Python 3.10

drGrove committed:
Remove python 3.7 and 3.8, add coveralls done notification

drGrove committed:
Add missing logging overrides to SQLLite tests

drGrove committed:
Combine coverage prior to running report

drGrove committed:
Allow coveralls combine to fail

drGrove committed:
Use curl container

drGrove committed:
Ensure that push doesn't run on PRs where it shouldn't happen

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,93 @@
+            return error_response("authentication required", 401)
+
+        if token_type == "PGP-SIG":
+            data = request.get_data()
+
+            # Move the pointer to the input stream back to 0 so that this can be accessed
+            # by the actual request
+            request.input_stream.seek(0)
```
Wait..... so if the endpoint has "login required", then you pass a PGP signature as the GET/POST body?
I'm not a fan of that design

drGrove replied:
No, it is passed as a Authorization header and is b64encoded. The reason there is a seek in that call is because we need to be able to verify the content of the body with the detached signature.

The original design had the signature in the body and generally only signed part of the data. This now signs all of the data but does have the fun problem that all fields in the json object must now be alphabetized to to allow for verification since json.loads has a sort_keys attribute and does so by default in flask apparently

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,93 @@
+        if not request.headers['Authorization']:
+            return error_response("authentication required", 401)
+
+        token_type, token = request.headers['Authorization'].split(' ')
```
What does this do if there is more than one space character in the header?

james-callahan replied:
Might be worth using the walrus here?
```suggestion
        if not (authorization_header := request.headers['Authorization']):
            return error_response("authentication required", 401)

        token_type, token = authorization_header.split(' ')
```

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,93 @@
+            return error_response("authentication required", 401)
+
+        if token_type == "PGP-SIG":
+            data = request.get_data()
```
From docs:
> Usually it’s a bad idea to call this method without checking the content length first as a client could send dozens of megabytes or more to cause memory problems on the server.

What are you doing to protect against that?
I also notice there is no timeout: would this be vulnerable to a slow loris attack?

drGrove replied:
There's a config we can pass through, I saw that in the docs and I there was some default but I may have also looked at the nginx docs to see if there is a default there (which there is, it's 1Mb). But then we'd be making the assumption that someone is using nginx or another proxy in front of this. 

Good catch

drGrove replied:
I actually did think of this: https://github.com/drGrove/mtls-server/pull/56/files#diff-8a171bbdfde7635fb9f28f0fd9840c48b1800d471ab36b64fe9714ee72a37249R41

I just didn't make it a default configuration in the kustomize

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,93 @@
+                return error_response("unauthorized", 401)
+
+            now = time.time()
+            if not time_in_range(now-5, now, int(verified.timestamp)):
```
This should probably be configurable rather than a hard-coded 5?
At least for now just hoist out to be a constant.

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,93 @@
+    @wraps(f)
+    def login_required_wrap(*args, **kwargs):
+        allowed_tokens = ['PGP-SIG']
+        g.is_admin = False
```
TIL flask `g` variable magic

drGrove committed:
Clean up authorization header assignment

drGrove committed:
Configurable timeout for PGP signed messages.

Allow an operator to set the timeout on PGP Signed Messages.

This defaults to 5 seconds, which is generally more than enough time for a
request to be signed and then sent to the server to begin processing. Allowing
this to be configurable will allow operators to teak the timeout based on
real-world use within their infrastructure.

drGrove committed:
Add test and try/catch to splitting authorization header

drGrove committed:
Remove unused import in test_sync

drGrove committed:
Remove unused verification code in cert_processor

drGrove committed:
Add test for seeding binary keys

drGrove committed:
Move away from openssl RSA key as we have in the server

drGrove committed:
Add type hinting to prevent python warnings

drGrove committed:
Clean up CRL creation for test

drGrove committed:
Add constant string for signing requests with no body content

drGrove committed:
Normalize folder creation

drGrove committed:
Better logging and commenting around specific edgecases

drGrove committed:
Use correct Config

drGrove committed:
Reference g.json instead of request.get_json() as the data has already been pull because of the auth check

james-callahan commented on integration_test/test_users.py:
```diff
@@ -180,19 +180,15 @@ def add_admin_invalid_admin(self):
-        }
         sig = self.admin_gpg.sign(
-            json.dumps(payload, sort_keys=True),
+            "NOCONTENT",
```
- What is this string for?
  - Why aren't you e.g. just signing the empty string?
  - Doesn't this require client changes?

drGrove replied:
Everything in this PR requires client changes.

When trying to sign empty content gpg was complaining. I had considered dropping the signing of content all together and moving to a more JWT like layout where the with header would have a base64 encoded string of the signing fingerprint along with a base64 encoded detached signature but didn't want to make more changes

drGrove replied:
The string exists to have some piece of data to sign for the request. On the server we check the signature as a form of auth. To prevent replays we only allow something signed within a few seconds of the request

drGrove committed:
Backport support for legacy certificate creation until clients can recommend to end users to have their servers upgraded

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,137 @@
+        allowed_tokens = ['PGP-SIG']
+        g.is_admin = False
+
+        if not (authorization_header := request.headers['Authorization']):
```
no need for walrus operator when you're using it outside of the `if` condition
```suggestion
        authorization_header := request.headers['Authorization']
        if not authorization_header:
```

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,137 @@
+
+        if token_type == "PGP-SIG":
+            if request.content_length:
+                g.data = data = request.get_data()
```
TIL this is valid python

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,137 @@
+                logger.debug(f"authenticated user {verified.pubkey_fingerprint} is admin")
+                g.is_admin = True
+
+            if not g.is_admin:
```
```suggestion
            else:
```

james-callahan commented on mtls_server/auth.py:
```diff
@@ -0,0 +1,137 @@
+            )
+            if verified.trust_level is not None and verified.trust_level >= verified.TRUST_ULTIMATE:
+                logger.debug(f"authenticated user {verified.pubkey_fingerprint} is admin")
+                g.is_admin = True
```
```suggestion
            g.is_admin = verified.trust_level is not None and verified.trust_level >= verified.TRUST_ULTIMATE
            if g.is_admin:
                logger.debug(f"authenticated user {verified.pubkey_fingerprint} is admin")
```

james-callahan commented on mtls_server/storage.py:
```diff
@@ -147,26 +147,27 @@ def get_cert(
+        query_options.append("revoked=?")
         if show_revoked:
-            query += " AND revoked=1"
+            values.append(str(1))
```
Why isn't this just
```suggestion
            values.append('1')
```

drGrove committed:
Simplify code paths in auth and getting revoked certs

Instead of checking the state of the verification object then setting the user
as admin, use the check to set True/False.

For setting weather or not to get revoked certificates instead of casting a
nubmer to string just use the string value

james-callahan commented on mtls_server/server.py:
```diff
@@ -1,69 +1,213 @@
+from .auth import login_required
+from .auth import legacy_verify
+from .cert_processor import CertProcessor
 from .cert_processor import CertProcessorKeyNotFoundError
+from .cert_processor import CertProcessorKeyNotFoundError
```
duplicated.

james-callahan commented on mtls_server/utils.py:
```diff
@@ -185,3 +195,31 @@ def import_and_trust(key_data, gpg):
+    if start <= end:
+        return start <= t <= end
+    else:
+        return start <= t or t <= end
```
shouldn't this whole `else` be a false? or an error?

drGrove committed:
Remove duplicate import

drGrove committed:
Simplify time_in_range and add test
